src/Covenant/Type.hs

```diff
@@ -706,3 +713,132 @@ substitute index toSub = \case
   ThunkT (CompT abstractions xs) ->
     ThunkT . CompT abstractions . fmap (substitute index toSub) $ xs
   BuiltinFlat t -> BuiltinFlat t
+
+-- Pretty printing stuff, somewhat awkward to put it here but doing so avoids orphan instances
+
+-- Keeping the field names for clarify even if we don't use them
+newtype ScopeBoundary = ScopeBoundary {_getBoundary :: Int}
+  deriving newtype (Show, Eq, Ord, Num, Enum)
+
+data PrettyContext (ann :: Type)
+  = PrettyContext {
+      _pcBoundIdents :: Map ScopeBoundary [Doc ann]
+    , _pcCurrentScope :: ScopeBoundary
+    , _pcVarStream :: [Doc ann]
+    }
+
+-- Lazily generated infinite list of variables. Will start with a, b, c... and cycle around to a1, b2, c3 etc.
+-- We could do something more sophisticated but this should work.
+infiniteVars :: forall (ann :: Type). [Doc ann]
+infiniteVars = let aToZ = ['a'..'z']
+                   intStrings = ("" <$ aToZ) <> map (show @Integer) [0..]
+               in zipWith (\x xs -> pretty (x:xs)) aToZ intStrings
+
+-- optics, for convenience. Manually defined b/c TH splices tend to blow up HLS
+boundIdents :: forall (ann :: Type). Lens' (PrettyContext ann) (Map ScopeBoundary [Doc ann])
+boundIdents = lens goGet goSet
+  where
+    goGet (PrettyContext bi _ _) = bi
+    goSet (PrettyContext _ scop vars) bi' = PrettyContext bi' scop vars
+
+currentScope :: forall (ann :: Type). Lens' (PrettyContext ann) ScopeBoundary
+currentScope = lens goGet goSet
+  where
+    goGet (PrettyContext _ scop _) = scop
+    goSet (PrettyContext bi _ vars) scop  = PrettyContext bi scop vars
+
+varStream :: forall (ann :: Type). Lens' (PrettyContext ann) [Doc ann]
+varStream = lens goGet goSet
+  where
+    goGet (PrettyContext _ _ vars) = vars
+    goSet (PrettyContext bi scop _) = PrettyContext bi scop
+
+-- Generate N fresh var names and use the supplied monadic function to do something with them.
+withFreshVarNames :: forall (ann :: Type) (a :: Type). Int -> ([Doc ann] -> PrettyM ann a) -> PrettyM ann a
+withFreshVarNames n act = do
+  stream <- asks (view varStream)
+  let (used,rest) = splitAt n stream
+  local (set varStream rest) $ act used
+
+-- Maybe make a newtype with error reporting since this can fail, but do later since *should't* fail
+newtype PrettyM (ann :: Type) (a :: Type) = PrettyM (Reader (PrettyContext ann) a)
+  deriving newtype (Functor,Applicative,Monad,MonadReader (PrettyContext ann))
+
+runPrettyM :: forall (ann :: Type) (a :: Type). PrettyM ann a -> a
+runPrettyM (PrettyM ma) = runReader ma (PrettyContext mempty 0 infiniteVars)
+
+
+-- REVIEW @Koz: You said we always consider a boundary crossed even if the count is zero,
+--              but when I do that it blows up lists? I think I have to be misunderstanding something
```

Comment:
Is this still current?

Reply:
I guess not, since we removed the builtin list type. I guess if it happens again with the new representation of lists I'll address it then.